replication/AmbryReplication.tla

```diff
@@ -0,0 +1,328 @@
+    inflightReplicationMessages, \* Inflight replication messages
+    inflightReplicationRequest, \* Inflight request by a source replica to get next set of replication data from target replica.
+    inflightReplicationResponse, \* Inflight response to replication sent by a replica to the replication requester.
+
+    (* The variables below are not system state but are needed to check that all user operations are done.
+     * Once all the operations are done, and each replica has had a chance to do replication atleast once, then we can check for consistency.
+     *)
+    deleteCtr, \* Count of delete operations.
+    updateTtlCtr, \* Count of update ttl operations.
+    recvReplicationReqSet, \* Set of replicas that executed RecvReplicationRequest.
+    sendRecplicationSet, \* Set of replicas that executed SendReplicationRequest.
+    recvReplicationRespSet \* Set of replicas that executed RecvReplicationResponse.
+
+vars == <<replicaStates,
+    storedMessagesPerReplica,
+    blobIdCtr,
+    inflightReplicationMessages,
+    inflightReplicationRequest,
+    inflightReplicationResponse, 
+    deleteCtr,
+    updateTtlCtr,
+    recvReplicationReqSet,
+    sendRecplicationSet,
+    recvReplicationRespSet>>
+
+StoredMessagesSet ==
+    { [ blobid |-> i, blobmessage |-> m ]
+        : i \in BlobIds, m \in {DELETE, UPDATE_TTL} }
+
+BlobCountConstraint ==
+    /\ blobIdCtr <= NumBlobs
+
+(* Replica storage record for a PUT message *)
+PutMessage(message) ==
+    [ blobid |-> message.blobid, blobmessage |-> PUT ]
+
+(* Replica storage record for a DELETE message *)
+DeleteMessage(message) ==
+    [ blobid |-> message.blobid, blobmessage |-> DELETE ]
+
+(* Replica storage record for a UPDATE TTL message *)
+UpdateTtlMessage(message) ==
+    [ blobid |-> message.blobid, blobmessage |-> UPDATE_TTL ]
+
+TypeOK ==
+    /\ replicaStates \in [Replicas -> ReplicaStateSet]
+    /\ DOMAIN inflightReplicationMessages \subseteq Replicas
+    /\ DOMAIN storedMessagesPerReplica \subseteq Replicas
+    /\ DOMAIN inflightReplicationRequest \subseteq Replicas
+    /\ DOMAIN inflightReplicationResponse \subseteq Replicas
+    /\ recvReplicationReqSet \subseteq Replicas
+    /\ sendRecplicationSet \subseteq Replicas
+    /\ recvReplicationRespSet \subseteq Replicas
+
+
+Init ==
+    /\ storedMessagesPerReplica = [ r \in Replicas |-> {} ]
+    /\ inflightReplicationMessages = [ r \in Replicas |-> {} ]
+    /\ inflightReplicationRequest = [ r \in Replicas |-> {} ]
+    /\ inflightReplicationResponse = [ r \in Replicas |-> {} ]
+    /\ replicaStates = [ r \in Replicas |-> CanTakeWrite ]
+    /\ blobIdCtr = 0
+    /\ deleteCtr = 0
+    /\ updateTtlCtr = 0
+    /\ recvReplicationReqSet = {}
+    /\ sendRecplicationSet = {}
+    /\ recvReplicationRespSet = {}
+
+(* If each blob has seen PUTs, updateTtl and Delete then we can assume that all user operations have ended. *)
+AllBlobOpsDone == 
```

Comment:
Conjunction or disjunction?  Is the model meant to show that all blobs get created/updated/deleted in a run?